Gradle fails with cryptic error when `plugins {}` contains unexpected code

Can't run JUnit tests with GradleRunner (tested on 3.5 and 4.10.2)

Minimal test-case from [gradle-node-plugin:](https://github.com/node-gradle/gradle-node-plugin) https://gist.github.com/deepy/b53c1bcc1c4f92c21ac6f182db413b71

(also happening on travis)
https://travis-ci.org/node-gradle/gradle-node-plugin/jobs/462819188#L656
```
    A problem occurred configuring root project 'junit2647550049405801039'.
    > Could not open cp_proj remapped class cache for aymfmvqz7m4rx0xrpgrjhp46f (C:\Users\deepy\AppData\Local\Temp\.gradle-test-kit-deepy\caches\3.5\scripts-remapped\build_3hr5fzwun61jnjy0zqxoam0ig\aymfmvqz7m4rx0xrpgrjhp46f\cp_proj0341ca279e16587311ec2a53cd62afc0).
       > Could not open cp_proj generic class cache for build file 'C:\Users\deepy\AppData\Local\Temp\junit2647550049405801039\build.gradle' (C:\Users\deepy\AppData\Local\Temp\.gradle-test-kit-deepy\caches\3.5\scripts\aymfmvqz7m4rx0xrpgrjhp46f\cp_proj\cp_proj0341ca279e16587311ec2a53cd62afc0).
          > BUG! exception in phase 'conversion' in source unit '_BuildScript_' org.codehaus.groovy.ast.expr.TupleExpression cannot be cast to org.codehaus.groovy.ast.expr.ArgumentListExpression
```

### Steps to Reproduce (for bugs)
Run any of the _integTest classes (e.g. Node_integTest.groovy)
```
./gradlew :test --tests "com.moowork.gradle.node.task.Node_integTest.exec simple node program"
```

### Your Environment
https://github.com/node-gradle/gradle-node-plugin
https://travis-ci.org/node-gradle/gradle-node-plugin/jobs/462819188#L656

This is not a bug in the Groovy compiler, but rather a bug in the Gradle AST transformation that extracts plugin usages from the `plugins` block. The underlying error is:

```
Caused by: java.lang.ClassCastException: org.codehaus.groovy.ast.expr.TupleExpression cannot be cast to org.codehaus.groovy.ast.expr.ArgumentListExpression
        at org.gradle.groovy.scripts.internal.AstUtils.hasSingleConstantArgOfType(AstUtils.java:221)
        at org.gradle.plugin.use.internal.PluginUseScriptBlockMetadataExtractor$1.visitMethodCallExpression(PluginUseScriptBlockMetadataExtractor.java:118)
        at org.codehaus.groovy.ast.expr.MethodCallExpression.visit(MethodCallExpression.java:70)
        at org.gradle.plugin.use.internal.PluginUseScriptBlockMetadataExtractor$1.visitExpressionStatement(PluginUseScriptBlockMetadataExtractor.java:155)
        at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:42)
        at org.gradle.plugin.use.internal.PluginUseScriptBlockMetadataExtractor$1.visitBlockStatement(PluginUseScriptBlockMetadataExtractor.java:63)
        at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:71)
        at org.gradle.plugin.use.internal.PluginUseScriptBlockMetadataExtractor.extract(PluginUseScriptBlockMetadataExtractor.java:58)
        at org.gradle.groovy.scripts.internal.InitialPassStatementTransformer.transform(InitialPassStatementTransformer.java:73)
        at org.gradle.groovy.scripts.internal.AstUtils.filterAndTransformStatements(AstUtils.java:90)
        at org.gradle.groovy.scripts.internal.SubsetScriptTransformer.call(SubsetScriptTransformer.java:51)
        at org.codehaus.groovy.control.CompilationUnit.applyToSourceUnits(CompilationUnit.java:976)
```

This is caused by unexpected code in the `plugins` block:

```
plugins {
   apply plugin: 'com.github.node-gradle.node'
}
```

Which should be:

```
plugins {
   id "com.github.node-gradle.node" version "1.2.0"
}
```

--- a/subprojects/base-services-groovy/src/main/java/org/gradle/groovy/scripts/internal/AstUtils.java
+++ b/subprojects/base-services-groovy/src/main/java/org/gradle/groovy/scripts/internal/AstUtils.java
@@ -220,13 +220,16 @@ public abstract class AstUtils {
 
     @Nullable
     public static ConstantExpression hasSingleConstantArgOfType(MethodCallExpression call, Class<?> type) {
-        ArgumentListExpression argumentList = (ArgumentListExpression) call.getArguments();
-        if (argumentList.getExpressions().size() == 1) {
-            Expression argumentExpression = argumentList.getExpressions().get(0);
-            if (argumentExpression instanceof ConstantExpression) {
-                ConstantExpression constantArgumentExpression = (ConstantExpression) argumentExpression;
-                if (isOfType(constantArgumentExpression, type)) {
-                    return constantArgumentExpression;
+        Expression arguments = call.getArguments();
+        if (arguments instanceof ArgumentListExpression) {
+            ArgumentListExpression argumentList = (ArgumentListExpression) arguments;
+            if (argumentList.getExpressions().size() == 1) {
+                Expression argumentExpression = argumentList.getExpressions().get(0);
+                if (argumentExpression instanceof ConstantExpression) {
+                    ConstantExpression constantArgumentExpression = (ConstantExpression) argumentExpression;
+                    if (isOfType(constantArgumentExpression, type)) {
+                        return constantArgumentExpression;
+                    }
                 }
             }
         }

--- a/subprojects/core/src/main/java/org/gradle/plugin/use/internal/PluginUseScriptBlockMetadataCompiler.java
+++ b/subprojects/core/src/main/java/org/gradle/plugin/use/internal/PluginUseScriptBlockMetadataCompiler.java
@@ -34,6 +34,7 @@ import org.gradle.groovy.scripts.internal.RestrictiveCodeVisitor;
 import org.gradle.groovy.scripts.internal.ScriptBlock;
 
 import static org.gradle.groovy.scripts.internal.AstUtils.hasSingleConstantArgOfType;
+import static org.gradle.groovy.scripts.internal.AstUtils.hasSingleConstantStringArg;
 import static org.gradle.groovy.scripts.internal.AstUtils.hasSinglePropertyExpressionArgument;
 import static org.gradle.groovy.scripts.internal.AstUtils.isOfType;
 
@@ -99,7 +100,7 @@ public class PluginUseScriptBlockMetadataCompiler {
                                 restrict(call, formatErrorMessage(DISALLOWED_ALIAS_NOTATION));
                                 return;
                             case "id":
-                                ConstantExpression argumentExpression = hasSingleConstantArgOfType(call, String.class);
+                                ConstantExpression argumentExpression = hasSingleConstantStringArg(call);
                                 if (argumentExpression == null) {
                                     restrict(call, formatErrorMessage(NEED_LITERAL_STRING));
                                     return;
@@ -170,7 +171,7 @@ public class PluginUseScriptBlockMetadataCompiler {
      * b) A GString expression containing only variable expressions
      */
     private static boolean hasSimpleInterpolatedStringType(MethodCallExpression call) {
-        if (hasSingleConstantArgOfType(call, String.class) != null) {
+        if (hasSingleConstantStringArg(call) != null) {
             return true;
         }
 

--- a/subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/PluginUseDslIntegrationSpec.groovy
+++ b/subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/PluginUseDslIntegrationSpec.groovy
@@ -200,6 +200,7 @@ class PluginUseDslIntegrationSpec extends AbstractIntegrationSpec {
         2          | "apply false"                          | BASE_MESSAGE
         2          | "id 'foo' apply"                       | BASE_MESSAGE
         2          | "id 'foo' apply('foo')"                | NEED_SINGLE_BOOLEAN
+        2          | "apply plugin: 'java'"                 | NEED_SINGLE_BOOLEAN
         2          | "id null"                              | NEED_LITERAL_STRING
         2          | "id 'foo' version null"                | NEED_INTERPOLATED_STRING
         2          | "file('foo')" /* script api */         | BASE_MESSAGE

